### ISSUE TITLE
Indentation and Comments inconsistent with the original code

### ISSUE THREAD

**Issue by Symbolk** [guilhermejccavalcanti/jFSTMerge#42]
Since the code is parsed and prettyprinted with JavaCC, though indented with JavaParser, the indentation of the original code and some comments (it's strange that some are orphan comments but some are not) are lost. When diff with git merged results using textual tools (like git diff or VSCode diff), there will be too many differences. And in some cases, comments are as important as codes.

I am trying to preserve the indentation and comments in original files, but in face of some strange problems. I want to know as authors or contributors, what's your way of thinking about implementing it? @guilhermejccavalcanti @pauloborba @gaabs

**Comment by pauloborba** [comment]
@leoribeiro36 this might be the explanation for some of the differences that lead to s3m FNs. I would not expect many FNs due to that, but maybe a few.

**Comment by pauloborba** [comment]
@Symbolk I guess the basic idea would be to keep comments and indentation info in the AST (as in Eclipse JDT), but I haven't really thought in detail about that. Could you please detail a bit more the problems you are having? Have you also seen this alternative approach?  https://www.modelica.org/events/modelica2008/Proceedings/sessions/session6a3.pdf

**Comment by pauloborba** [comment]
@jvcoutinho could you please add tests to check how the tool should handle comments? please check @Symbolk comments above, and ask for the concrete merge scenarios in which he observed lost comments.

**Comment by jvcoutinho** [comment]
@Symbolk as requested above by @phmb, can you please post here the files that lead to outputs losing comments? You can compact them.

**Comment by jvcoutinho** [comment]
![s3m](https://user-images.githubusercontent.com/27645414/46618538-29643a00-caf6-11e8-9e50-2c5a96440d2a.png)
![git](https://user-images.githubusercontent.com/27645414/46618539-29fcd080-caf6-11e8-9de2-560f1bff0d90.png)

Comments problem example

**Comment by pauloborba** [comment]
@jvcoutinho could not understand that. both developers commented different declarations? what's the merge result?

**Comment by jvcoutinho** [comment]
@pauloborba Actually, by their description, one developer included a field while another developer commented an existing one.

**Comment by jvcoutinho** [comment]
The only way I could reproduce this exact scenario is when the field in `base` is also commented (and it seems to be the case in the example):

![comment3](https://user-images.githubusercontent.com/27645414/46683682-2dac5800-cbc7-11e8-8d84-476a4e3b0396.png)
![comment3output](https://user-images.githubusercontent.com/27645414/46683681-2dac5800-cbc7-11e8-8b7e-895bcb35f177.png)

But that's reasonable, since only `right` changed the file.

**Comment by jvcoutinho** [comment]
Although, I'd tested more and found some inconsistencies, indeed.

![comment](https://user-images.githubusercontent.com/27645414/46683887-baefac80-cbc7-11e8-9f0a-688550024e0b.png)
![commentoutput](https://user-images.githubusercontent.com/27645414/46683890-bb884300-cbc7-11e8-845d-ce1ca36ab29b.png)

![comment2](https://user-images.githubusercontent.com/27645414/46683888-bb884300-cbc7-11e8-8c95-82441af7e71d.png)
![comment2output](https://user-images.githubusercontent.com/27645414/46683889-bb884300-cbc7-11e8-9120-ab22c5be23c8.png)

**Comment by pauloborba** [comment]
yes @jvcoutinho, we need to debug that. please check when base has field A, left removes A, right adds a comment just before A. the result should contain only the comment. does the tool returns an empty class?

**Comment by jvcoutinho** [comment]
yes, @pauloborba 
![comment4](https://user-images.githubusercontent.com/27645414/46692749-71f72280-cbde-11e8-8f74-495fe0eb07dc.png)
![comment4output](https://user-images.githubusercontent.com/27645414/46692748-71f72280-cbde-11e8-9a6a-520f0190fbb4.png)

**Comment by pauloborba** [comment]
OK @jvcoutinho, so we have enough information now. I believe s3m stores comments in the closest declaration node, not in special CommentNodes, right @guilhermejccavalcanti?

**Comment by guilhermejccavalcanti** [comment]
Right, comments are stored into a separate field ```prefix``` of each node:
![image](https://user-images.githubusercontent.com/3333981/46701482-2f414480-cbf6-11e8-9606-efa5bb578d9c.png)

Also, they are not merged. If the three versions have comments, the left comment prevails due to precedence order of the merge calls and how the merged node is created. First, ```merge(left,base)```, then ```merge(leftBase, right)```, each call creates the merged node as a clone of the first parameter, which is always left, and update its body/subtree with the merge result. So, besides updating node body/subtree, would be necessary to update the ```prefix``` field, perhaps calling textual merge on it.

**Comment by Symbolk** [comment]
Sorry for the delay to reply. 

Originally, I want to develope a merge tool based on jFSTMerge, which produces better or more reasonable merge results than git textual merge. 

However, since the code is parsed, merged and then prettyprinted with AST, the original indentation and orphan comments can be missing in the merged results, which are undesirable in practical projects. 

For example, in some big teams or projects (say AOSP), developers are required to wrap their newly added or edited code within two block comments, for subsequent review, blame or debug:

```java
import android.widget.Toolbar;

/* Peter/62645 20180637  begin*/
//[HSM]
import android.view.Menu;
/* Peter/62645 20180637  end*/
import com.android.internal.annotations.GuardedBy;

```
(P.S. the comment format: /* Developer/ID Date begin(end) */)

Under this case, the jFSTMerge results drops the end comment while the git merge results does not (of course). Here are the 3-way contents: 

Ours: 
```java
import android.widget.Toolbar;

/* Peter/62645 20180637  begin*/
//[HSM]
import android.view.Menu;
/* Peter/62645 20180637  end*/
import com.android.internal.annotations.GuardedBy;

```

Base: 
```java
import android.widget.Toolbar;

import com.android.internal.annotations.GuardedBy;
```

Theirs:
```java
import android.widget.Toolbar;

import com.android.internal.annotations.GuardedBy;
```

jFSTMerge results:
```java
import android.widget.Toolbar;

/* <DTS2014042818262 xiongshiyi/00165767 20140428 begin */
//[HSM]
import android.hsm.HwSystemManager;

import com.android.internal.annotations.GuardedBy;
```

**Comment by Symbolk** [comment]
> yes, @pauloborba
> ![comment4](https://user-images.githubusercontent.com/27645414/46692749-71f72280-cbde-11e8-8f74-495fe0eb07dc.png)
> ![comment4output](https://user-images.githubusercontent.com/27645414/46692748-71f72280-cbde-11e8-9a6a-520f0190fbb4.png)

What is the 3-way diff tool you are using? Seems VSCode?

**Comment by jvcoutinho** [comment]
@Symbolk Yes, it is VSCode.
You can split the editor 2 times and activate Zen Mode (Ctrl + K + Z by default) to get this view.

**Comment by Symbolk** [comment]
> @Symbolk Yes, it is VSCode.
> You can split the editor 2 times and activate Zen Mode (Ctrl + K + Z by default) to get this view.

Ok, thanks. Previously I thought that was an extension to provide the view like BeyondCompare 3-way merge, I am thinking about developing one like this.

**Comment by pauloborba** [comment]
that would be great @Symbolk. you might want to have a look at https://semanticmerge.com.

**Comment by pauloborba** [comment]
@jvcoutinho it seems we have similar problems with annotations too

**Comment by guilhermejccavalcanti** [comment]
Annotations are considered as part of declarations body, and as such are
merged with textual merged.

So the question is: annotations order matters? If so, the behavior is
correct. Otherwise, would be necessary to have separate nodes representing
annotations.

Em qui, 11 de out de 2018 14:29, Paulo Borba <notifications@github.com>
escreveu:

**Comment by pauloborba** [comment]
@jvcoutinho ask if ISI developers adopt an indentation tool/standard

**Comment by jvcoutinho** [comment]
@pauloborba this might be the next to go

**Comment by pauloborba** [comment]
yes @jvcoutinho, please go ahead

**Comment by pauloborba** [comment]
> that would be great @Symbolk. you might want to have a look at https://semanticmerge.com.

have you made progress here @Symbolk?

**Comment by jvcoutinho** [comment]
I've been looking the way the creator of [Spoofax](http://www.metaborg.org/en/latest/index.html), [Eelco Visser](https://eelcovisser.org/), had treated parsing and pretty-printing. In his paper "Declarative Specification of Indentation Rules", he says that the grammar should have special declarations to specify the indentation. These declarations are ignored by the parser (for layout-sensitive languages like Haskell or Python, these declarations play an important role in parsing too) but relevant for the pretty-printer.

**Comment by guilhermejccavalcanti** [comment]
>
> he says that the grammar should have special declarations to specify the
> indentation
>

So, the starting point is the generator package
https://github.com/guilhermejccavalcanti/jFSTMerge/tree/master/src/main/java/br/ufpe/cin/generator
which uses classes from the featurehouse.jar dependency. Afterwards, it
would be necessary to change the classes from this package to support new
annotations.

Em sex, 28 de jun de 2019 às 19:14, João Victor <notifications@github.com>
escreveu:

**Comment by pauloborba** [comment]
@jvcoutinho you could first try to solve the comments issue. 

this should be simpler, and satisfactory for companies that use a code formatting standard/tool. s3m could simply merge and then call the formatting tool.

then you try to solve the indentation/white space issue.

**Comment by jvcoutinho** [comment]
So, the comments issue.
Currently, S3M stores comments in the FSTTerminal field `prefix`. It contains all the tokens since the end of the last terminal read. This implies comments depend on terminals to be considered. Also, textual merge runs over this field, leading possibly to conflicts.

There are two main problems with the current approach:
- False positives;
- Orphan comments.

In the scenario below, `left removed method a()` and `right added a comment above method a() and another below a()`.

![Merge](https://user-images.githubusercontent.com/27645414/60912589-14226480-a25c-11e9-94dc-ffb5d3d06173.PNG)

Comments represent no semantic "threat". I believe comments merge shouldn't lead to conflicts (S3M used to consider left version, while SemanticMerge always consider right version). Anyway, resolving false positives is the easy part.

Orphan comments, on the other hand, are harder to deal with. I propose two solutions:

- Use unstructured merge to search for comments that s3m had forgotten. This would be basically a textual approach and it's more susceptible to errors.
- Store comments as tree nodes (just like JDT parser does). In this approach, comments have no identifiers, so we would have to build a handler for comments.

What do you think, @pauloborba @guilhermejccavalcanti?

**Comment by jvcoutinho** [comment]
1. Check if it's worth to update the grammar to include an attribute `comments` in a FSTNonTerminal node.
2. Check if we should keep a comment of a removed node.

3. S3M doesn't ignore comments in the middle of the signature:
![comment](https://user-images.githubusercontent.com/27645414/60998047-9b400d00-a32e-11e9-86e6-c577812b2288.PNG)

**Comment by pauloborba** [comment]
check example “comment public final…” vs “final comment static"

**Comment by jvcoutinho** [comment]
S3M reports 2 conflicts (the prefix field and the declaration field) in different lines.

![comment](https://user-images.githubusercontent.com/27645414/60998619-ac3d4e00-a32f-11e9-9043-9a898e7d8501.PNG)

**Comment by jvcoutinho** [comment]
**Indentação**
preservar original;
se nós forem adicionados ou removidos? (não há problema)
se left/right adicionarem exatamente o mesmo método, qual escolher?
espaçamento entre métodos, contar linhas ou padronizar?

**Comentários**
fazer com que a chave que fecha a classe ("}") virasse um FSTTerminal para capturar os comentários órfãos como prefixo

**Comment by guilhermejccavalcanti** [comment]
@jvcoutinho me deparei com um possível bug no cenário em anexo, dá uma checada por favor.
Basicamente, em left e base havia Comentário seguido de Anotação. E right inverteu isso, ficou Anotação seguida de Comentário.  Left não mudou nada no comentário nem na anotação, e deu conflito.

[comment.zip](https://github.com/guilhermejccavalcanti/jFSTMerge/files/3458142/comment.zip)

**Comment by jvcoutinho** [comment]
Bem, de fato é um bug. Vou colocar os trechos de código aqui para explicar.

**LEFT/BASE**
```
@LocalBean
@Stateless
/**
 * Classe de servido para gerenciar a entidade Apelido.
 * 
 * @author rbonifacio
 */
public class PalavraChaveServico extends ServicoGenerico<PalavraChave> {
```

**RIGHT**
```
/**
 * Classe de servido para gerenciar a entidade Apelido.
 * 
 * @author rbonifacio
 */
@LocalBean
@Stateless
public class PalavraChaveServico extends AbstractServicoSisdot<PalavraChave> {
```

**MERGE**
```
<<<<<<< MINE
=======
/**
 * Classe de servido para gerenciar a entidade Apelido.
 * 
 * @author rbonifacio
 */
>>>>>>> YOURS
@LocalBean
@Stateless
public class  PalavraChaveServico extends AbstractServicoSisdot<PalavraChave> {
```
Em left, tanto as anotações, o comentário e o token `public` foram colocados dentro do corpo de um mesmo nó (a princípio, esse nó conteria apenas os modificadores do método, o que sugere bug também na gramática). Em right, as anotações e o `public` foram colocados dentro do corpo de um mesmo nó. Porém, o comentário entrou como prefixo desse nó.

Assim, o conflito se deu quando o merge textual foi chamado nos prefixos. O estranho é que o prefixo de left é idêntico ao prefixo da base, então não sei ao certo o motivo do conflito. O não-estruturado não reportou.

Aí vai alguns prints:
Nó de left/base:
![2019-08-02 (3)](https://user-images.githubusercontent.com/27645414/62386894-86225c80-b52f-11e9-9bc0-c764c60bfe21.png)

Nó de right:
![2019-08-02 (4)](https://user-images.githubusercontent.com/27645414/62386988-c97ccb00-b52f-11e9-93d5-b6f605c5df4b.png)

**Comment by pauloborba** [comment]
@jvcoutinho please check if this is related to the case where left removes declaration but not its associated comment. add a test case for that. 

also check how modifiers are represented in the tree.

**Comment by jvcoutinho** [comment]
@pauloborba 

> also check how modifiers are represented in the tree.

Exactly one Modifiers node is created for each class declaration, and it contains all of the modifiers in its content.

> please check if this is related to the case where left removes declaration but not its associated comment.

the result is only the comment (and only one, no duplicates), but wrapped in a conflict:
![image](https://user-images.githubusercontent.com/27645414/62504088-65773280-b7cd-11e9-82f7-bb58c07db8dd.png)

although, just as the bug above, one of the contributions and the base are **exactly** the same. I believe prefix merge has a internal bug.

**Comment by pauloborba** [comment]
@jvcoutinho yes, it should be handled as a child of a non-terminal node, right?

**Comment by guilhermejccavalcanti** [comment]
I believe that the comments are not in the prefix because they are beetwen the annotations and the modifier. So, I guess this is the condition implemented on featurehouse "the comments are prefix only if there is nothing before them". Not sure if this is a grammar issue since this is based on original Java 8 grammar. We should avoid changes on grammar unless it is wrong, otherwise the tool will have behavior hard to explain.

Now, the question about "why is a conflict reported if there is only a single change?". This happens because the base version of the comment is empty, so, as there is no base, textual merge applies a two-way merge, and reports a conflict for every difference.

![image](https://user-images.githubusercontent.com/3333981/62560554-ae71ca00-b853-11e9-9256-19ae7c9ace1b.png)


So, a slightly improvement would be to compare the strings representing the three versions, and when only one of them is different, this is the difference is merge result. This will fix this issue.
This is actually what we do for the entire files (https://github.com/guilhermejccavalcanti/jFSTMerge/blob/master/src/main/java/br/ufpe/cin/app/JFSTMerge.java#L210)

**Comment by pauloborba** [comment]
@jvcoutinho would we have a similar problem when merging a method body that is empty in the base?

**Comment by jvcoutinho** [comment]
@pauloborba No because its signature is included in the content to be merged. Base could never be empty in these cases.

**Comment by jvcoutinho** [comment]
> Now, the question about "why is a conflict reported if there is only a single change?". This happens because the base version of the comment is empty, so, as there is no base, textual merge applies a two-way merge, and reports a conflict for every difference.

@guilhermejccavalcanti @pauloborba Fixed in the current PR (#89).

**Comment by jvcoutinho** [comment]
I'll try to explain here what I did for indentation (currently on #89), since it's a little complex.

Basically, the `prefix` field contains all the information I need.
I parse this field and divide it in 5 consecutive strings/fields:

**PreLineBreaks** ||  **PreIndentation** || **Comment** || **PostLineBreaks** || **PostIndentation**

**PreLineBreaks**: all the line breaks ('\n') between the last token and the comment.
**PreIndentation**: the comment indentation.
**Comment**: well, the comment. It can also contain more than one comment, but we can think as only one without lose of generality.
**PostLineBreaks**: all the line breaks between the comment and the node's content.
**PostIndentation**: the node's indentation. It's the only field mandatory (non-empty).

I ignore the PreLineBreaks group in order to put line breaks manually. So, in order to indent the nodes, I simply print the prefix followed by the body.

Although, there are two cases where this approach fails:

- Conflicts: the first line of a class member (method, field...) is never indented. This happens because the indentation is present only in the prefix attribute (PostIndentation). So, when S3M is generating a conflict, I use the PostIndentation to indent the first line of both contributions. This lead to a change of the class MergeConflict's constructors (and of all of the classes that instantiate it). Also, I have to remove the indentation from the prefix because it would incorrectly indent the first conflict marker.

- The node has been deleted but a comment remained: this can be achieved if a contribution added a comment above a node and the other deleted the node. I remove the PostLineBreaks and PostIndentation groups from the prefix, otherwise it would space the next node more than necessary.
 
https://github.com/guilhermejccavalcanti/jFSTMerge/blob/3c06e8f4532ffb86fe79c51597e7ae5e5f3e671e/src/main/java/br/ufpe/cin/printers/S3MPrettyPrinter.java#L44-L53

**Comment by pauloborba** [comment]
@jvcoutinho please add tests

**Comment by pauloborba** [comment]
@guilhermejccavalcanti could you please review that?

**Comment by pauloborba** [comment]
> @pauloborba No because its signature is included in the content to be merged. Base could never be empty in these cases.

just to make sure that a similar problem doesn't happen when handling bodies in isolation.
